# 7장. 오류 처리

### 목표

우아하게 예외를 처리하는 방법을 알아본다. 논리와 오류 처리를 분리하고, 의존성이 적은 오류 처리를 살펴본다.

### 목차

1. 예외 처리 방식 : 오류 코드 vs 예외 처리
2. Unchecked Exception을 사용하라
3. Exception 잘 사용하기
    1. 예외에 메시지 담기
    2. Exception Wrapper 사용하기
4. 실무 예외 처리 패턴
    1. null 처리
        - getOrElse : 예외 대신 기본값을 리턴한다
        - getOrElseThrow : null 대신 예외를 던진다
        - Optional
    2. 파라미터의 null 점검
    3. 자신의 예외 정의

## 💡예외 처리 방식

당연한 얘기지만, 오류 코드를 리턴하지 말고 예외를 던지는 방식을 사용해라.

<img width="80%" alt="Untitled" src="https://user-images.githubusercontent.com/53958188/174446139-85203271-3d0b-468e-99b8-196f40dd4827.png">

오류 코드를 사용하면 바로 오류를 처리해야 하므로 로직과 오류 처리 코드가 섞인다. 호출자 코드가 복잡해지고 의미를 파악하기 어렵다.

<img width="80%" alt="Untitled 1" src="https://user-images.githubusercontent.com/53958188/174446147-2bf88084-6e8b-4bbb-9823-42ff000eeaaa.png">

1. 오류가 발생한 부분에서 예외를 던진다. 별도의 처리가 필요한 예외라면 checked exception으로 던진다.
2. checked exception에 대한 예외 처리를 하지 않는다면 메서드 선언부에 throws를 명시해야한다.
3. 예외를 처리할 수 있는 곳에서 catch하여 처리한다.

## 💡Unchecked Exception을 사용하라

### Checked vs Unchecked Exception

<img width="80%" alt="Untitled 2" src="https://user-images.githubusercontent.com/53958188/174446162-431bc062-fe3c-452c-be25-671d7e4b3f13.png">

Checked Exception은 Exception을 직접 상속 받은 예외 클래스이다. IOException, SQLException 등등. 반면, Unchecked Exception는 Runtime Exception을 상속 받은 예외 클래스이다. 

NullPointerException, IllegalArgumentException, 
IndexOutOfBoundException 등등.

### Checked Exception이 나쁜 이유

<img width="80%" alt="Untitled 1" src="https://user-images.githubusercontent.com/53958188/174446184-564ccc4d-49c3-4211-bbb3-f2791df3300e.png">

Checked는 명시적인 예외 처리가 필요하다. 이 과정에서 OCP에 위배된다. 예외를 발생시킨 하위 메소드와 예외를 처리하는 상위 메소드의 **중간 단계의 메소드는 모두 특정 예외를 throw해야 한다.** 확인된 예외가 캡슐화를 깨버리며 의존성의 비용을 발생시키고 있다. 

### Unchecked Exception을 사용하자

Unchekced Exception은 명시적인 예외 처리가 필요하지 않다. 로직과 예외 처리를 분류하면서도 OCP를 지킬 수 있다.

![Untitled 3](https://user-images.githubusercontent.com/53958188/174446188-6dc03f94-ccbc-4427-a1a2-f935c87f85fa.png)

## 💡Exception 잘 사용하기

### 예외에 메시지 담기

<img width="80%" alt="Untitled 4" src="https://user-images.githubusercontent.com/53958188/174446200-a4c2720e-ed80-4616-9971-c3c9acfb433c.png">

애플리케이션이 로깅 기능을 사용한다면 catch 블록에서 오류를 기록하도록 충분한 정보를 넘겨준다.

### Exception Wrapper 사용하기

<img width="80%" alt="Untitled 5" src="https://user-images.githubusercontent.com/53958188/174446206-2913b1bb-74c3-4942-9368-2156672f0d86.png">

오류를 정의하는 기준은 발생 위치, 유형이 될 수 있지만 가장 중요한 관심사는 오류를 잡아내는 방법이 되어야 한다. (구체적으로는 catch의 로직)

왼쪽 코드는 여러 에러를 잡지만, 로그를 찍을 뿐 처리 방식에는 차이가 없다. 차라리 예외 클래스로 한번 감싸서 하나의 예외를 처리하는 방식으로 구현하는 것이 더 깔끔하다.

ACMEPort를 선언해 port.open()을 호출하고 여기서 발생할 수 있는 모든 예외를 다른 catch문을 잡았다. ACMEPort를 내부 필드로 가지는 LocalPort 클래스를 선언해 acmeport.open()이 발생시킬 수 있는 모든 예외에 대해 동일한 예외를 throw하도록 작성한다. 이제 localport.open()을 호출하는 곳에서는 하나의 예외만 catch로 처리하면 된다.

실제로 외부 API를 사용할 때는 감싸기 기법이 최선이다. 외부 API를 감싸면 외부 라이브러리와 프로그램 사이에 의존도가 크게 줄어든다. 특정 업체가 API를 설계한 방식에 의존적이지 않게 된다. 다른 라이브러리로 변경해도 비용이 작다. 감싸기 클래스에서 외부 API를 호출하는 대신 테스트 코드를 넣어주면 프로그램 테스트도 쉬워진다. 

## 💡실무 예외 처리 패턴

1. null 처리
    - getOrElse : 예외 대신 기본값을 리턴한다
    - getOrElseThrow : null 대신 예외를 던진다
    - Optional
2. 파라미터의 null 점검
3. 자신의 예외 정의

### null 처리

null을 반환하고 나서 if(object ≠ null)로 확인하는 방식은 나쁘다. null 체크를 빼먹을 수도 있고, 오류 처리와 로직이 뒤섞이기 때문이다. 그렇다면 null이 아닌 다른 값을 반환하는 방식과 예외를 던지는 방식을 살펴보자.

1. getOrElse (메소드 이름)
    
    **null이 아닌 기본값** 리턴하기. 예를 들어 데이터가 없음을 의미하는 컬렉션
    
    ```java
    List<Employee> employees = getEmployees();
    for(Employee e : employees) {
    	totalPay += e.getPay();
    }
    public List<Employee> getEmployees() {
    	if ( ...직원이 없다면... ) {
    		return **Collections.emptyList();**
    	}
    }
    ```
    
    **도메인에 맞는 기본값** 리턴하기. 예를 들어 기본값을 나타내는 코드
    
    ```java
    public UserService {
    		private static final UserLevel USER_BASIC_LEVEL = UserLevel.BASIC;
    	
    		public UserLevel getUserLevelOrDefault(Long userId){
    			try{
    				User user = userRepository.findByUserId(userId);
    				return user.getUserLevel();
    			} catch(UserNotFoundException e){
    				return **USER_BASIC_LEVEL**;
    			}
    		}
    }
    ```
    

1. getOrElseThrow  (메소드 이름)
    
    데이터를 제공하는 쪽에서 null 체크를 하여, 데이터가 없는 경우엔 null 대신 예외를던진다. 호출부에서 매번 null 체크를 할 필요 없이 안전하게 데이터를 사용할 수 있다. 호출부의 가독성이 올라간다.
    
    ```java
    public class UserService {
    	private static final UserLevel USER_BASIC_LEVEL = UserLevel.BASIC;
    	
    	public User **getUserOrElseThrow**(Long userId){
    		User user = userRepository.findByUserId(userId);
    		**if(user == null){
    			throw new IllegalArgumentException("User is not found. userId = "+userId);
    		}**
    		return user;
    	}
    }
    ```
    
2. Optional
    
    JpaRepository로 값을 조회하면 Optional에 감싸진 엔티티 객체를 받을 수 있다. Optional은 null이 올 수 있는 값을 감싸는 Wrapper 클래스로, 참조하더라도 NPE가 발생하지 않도록 도와준다. 올바른 사용법은 다음 [링크](https://www.latera.kr/blog/2019-07-02-effective-optional/) 참고. 
    
    ```java
    public Long getPrice(Long apartmentId){
    	return apartmentRepository.findById(apartmentId);
    			.orElseThrow(() -> new HoustUtilsException(ErrorCode.ENTITY_NOT_FOUND))
    			.getPrice();
    }
    ```
    

### 파라미터의 null 점검

null을 리턴하는 것도 나쁘지만 null을 메서드로 넘기는 것은 더 나쁘다. null을 메서드의 파라미터로 넣어야하는 API를 사용하는 경우가 아니면 null을 메서드로 넘기지마라.

```java
public class MetrixCalculator {
	public double xProjection(Point p1, Point p2){
		if(p1 == null || p2 == null){
			throw InvaludArgumentException("Invalid argument for MetrixCalculator.xProjection");
		}
		return (p2.x - p1.x) * 1.5;
	}
}
```

```java
public class MetrixCalculator {
	public double xProjection(Point p1, Point p2){
		assert p1 != null : "p1 should not be null";
		assert p2 != null : "p2 should not be null";
	}
}
```

### 자신의 예외 정의

```java
public class MyProjectException extends RuntimeException {
	private MyErrorCode errorCode;
	private String errorMessage;

	public MyProjectException(MyErrorCode errorCode){...}
	public MyProjectException(MyErrorCode errorCode, String errorMessage){...}
}

public enum MyErrorCode {
	private String defaultErrorMessage;
	
	INVALID_REQUEST("잘못된 요청입니다");
	DUPLICATED_REQUEST("기존 요청과 중복되어 처리할 수 없습니다.");
	INTERNAL_SERVER_ERROR("처리 중 에러가 발생했습니다.");
}

// 호출부
if(reuest.getUserName() == null){
	throw new **MyProjectException**(ErrorCode.INVALID_REQUEST, "user name is null");
}
```

에러 로그에서 stacktrace 해봤을 때 우리가 발생시킨 예외라는 것을 바로 인지할 수 있다. 다른 라이브러리에서 발생한 에러와 섞이지 않는다. 우리도 IllegalArgumentException을 던지는 것보다 우리 예외로 던지는게 어느 부분에서 에러가 났는지 파악하기에 용이하다. 우리 시스템에서 발생한 에러의 종류를 나열할 수 있다.
